sdk/azidentity/cache/linux.go

```diff
@@ -0,0 +1,242 @@
+//go:build go1.18 && linux
+// +build go1.18,linux
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package cache
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity/cache/internal/aescbc"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity/cache/internal/jwe"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity/internal"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/log"
+	"github.com/AzureAD/microsoft-authentication-extensions-for-go/cache/accessor"
+	"github.com/AzureAD/microsoft-authentication-extensions-for-go/cache/accessor/file"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	keySize = 32
+	userKey = "user"
+)
+
+var (
+	cacheDir   = os.UserHomeDir
+	tryKeyring = func() error {
+		k, err := newKeyring("azidentity-test-cache")
+		if err != nil {
+			return err
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cancel()
+		err = k.Write(ctx, []byte("test"))
+		if err != nil {
+			return err
+		}
+		_, err = k.Read(ctx)
+		if err != nil {
+			return err
+		}
+		return k.Delete(ctx)
+	}
+)
+
+func storage(o internal.TokenCachePersistenceOptions) (accessor.Accessor, error) {
+	name := o.Name
+	if name == "" {
+		name = defaultName
+	}
+	err := tryKeyring()
+	if err != nil {
+		msg := fmt.Sprintf("cache encryption is impossible because the key retention facility isn't usable: %s", err)
+		if o.AllowUnencryptedStorage {
+			f := ""
+			f, err = internal.CacheFilePath(name)
+			if err == nil {
+				log.Write(azidentity.EventAuthentication, msg+". Falling back to unencrypted storage")
+				return file.New(f)
+			}
+		} else {
+			err = errors.New(msg + ". Set AllowUnencryptedStorage on TokenCachePersistenceOptions to store the cache in plaintext instead of returning this error")
```

Comment:
Nit, maybe lower-case `Set` to be consistent with lower-casing?

Reply:
I dunno, this follows a period closing another sentence, shouldn't it be capitalized?

Reply:
True, and I waffled on commenting due to this.  I just think about how error messages in general aren't proper sentences, so the `S` looks weird to me.  Either way is fine though.

Reply:
> I just think about how error messages in general aren't proper sentences

A subject of much chagrin, IMO, especially when the fallback case seems to be "write the error to stderr (or stdout) and terminate". We could probably be more dev-friendly with actionable errors.